### PR TITLE
Reset module CQRS Command

### LINKED ISSUE #36979: Admin API - Reset module

### PR DESCRIPTION
**Pull request by jinpresta**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop 
| Description?      | Reset module CQRS Command
| Type?             | new feature  
| Category?         |  BO 
| BC breaks?        | ~
| Deprecations?     | ~
| How to test?      | Indicate how to verify that this change works as expected.
| UI Tests          |  https://github.com/jolelievre/ga.tests.ui.pr/actions/runs/11444800562
| Fixed issue or discussion?     | Fixes #36979 
| Related PRs       | ~
| Sponsor company   | ~

### CODE REVIEW

**Comment by ps-jarvis** [comment]
Hi, thanks for this contribution!

I found some issues with the Pull Request description:

* The `BC breaks` should be `yes` or `no`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#bc-breaks))
* The `deprecations` should be `yes` or `no`. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#deprecations))
* The `How to test` shouldn&#039;t be empty. ([Read explanation](https://devdocs.prestashop-project.org/9/contribute/contribution-guidelines/pull-requests/#how-to-test))

Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!
### About linked issues
Please consider opening an issue before submitting a Pull Request:
* If it's a bug fix, it helps maintainers verify that the bug is effectively due to a defect in the code, and that it hasn't been fixed already.
* It can help trigger a discussion about the best implementation path before a single line of code is written.
* It may lead the Core Product team to mark that issue as a priority, further attracting the maintainers' attention.

(Note: this is an automated message, but answering it will reach a real human)

**Comment by ps-jarvis** [comment]
Hello @jinpresta!

This is your first pull request on [PrestaShop](https://github.com/PrestaShop/PrestaShop/) repository of the PrestaShop project.

Thank you, and welcome to this Open Source community!

File: src/Adapter/Module/CommandHandler/ResetModuleHandler.php
Comment (jolelievre) [inline src/Adapter/Module/CommandHandler/ResetModuleHandler.php]: In the `GetModuleInfosHandler` we added a check to be sure the module is present, if not we return a 404 I think we should homogenize this behaviour in all actions So that if someone tries to do an action on a non-existent module he will always get a 404

To help doing that everywhere can you please move this code from the get handler:
```php
$module = $this->moduleRepository->getModule($query->getTechnicalName()->getValue());
        if (!$module->disk->get('is_present')) {
            throw new ModuleNotFoundException('Cannot find Module with name ' . $query->getTechnicalName()->getValue());
        }
```

into a `ModuleRepository::getPresentModule` method that performs this check and returns the `$module` at the end. This way n all handlers (including `GetModuleInfosHandler` and this handler) we can always start by doing:
```php
    $module = $this->moduleRepository->getPresentModule($technicalName);
```
and start working from this. This way we'll guaranty the not found use case and the 404 is correctly handled everywhere. And you can even use the returned instance to check if it's enabled or installed based on its attributes.

File: src/Core/Domain/Module/Command/ResetModuleCommand.php
Comment (jolelievre) [inline src/Core/Domain/Module/Command/ResetModuleCommand.php]:
```suggestion
    public function keepData(): bool
```

File: tests/Integration/Behaviour/Features/Scenario/Module/module.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Module/module.feature:56]: This step `Given module ps_featuredproducts has following infos:` is only used to check the status of the module, it only reads data so it won't modify the module state and should only be used to check the status after your reset action
We can't really test what the reset does but you can at least validate that it is still installed and enabled

```suggestion
    When I reset module "ps_featuredproducts"
    Then module ps_featuredproducts has following infos:
      | technical_name | ps_featuredproducts |
      | enabled        | true |
      | installed      | true |
```

File: tests/Integration/Behaviour/Features/Scenario/Module/module.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Module/module.feature]: I believe here you are testing that an exception is correctly thrown when you try to reset a module that is not installed or enabled, but this read step won't change the status into what you expect you need to use a write/update action to do that

 For now, we don't have a step to uninstall but we have one to disable
```suggestion
    When I disable module "ps_featuredproducts"
    And I reset module "ps_featuredproducts"
    Then I should have an exception that module is not enabled
```

File: tests/Integration/Behaviour/Features/Scenario/Module/module.feature
Comment (jolelievre) [inline tests/Integration/Behaviour/Features/Scenario/Module/module.feature]: You can remove the rest for now, in theory you should test that:
- you uninstall the module
- you try to reset
- you get an exception

But since we don't have the uninstall command yet this use case is not testable for the moment, you can leave a comment to remember to add it later when the uninstall command will be available

Comment (jolelievre) [reply tests/Integration/Behaviour/Features/Scenario/Module/module.feature]:
```suggestion
    # Todo: add the check on uninstalled module when the uninstall command is ready
    # Given I uninstall module "ps_featuredproducts"
    # When I reset module "ps_featuredproducts"
    # Then I should have an exception that module is not installed
```

**Comment by jolelievre** [comment]
One more thing you correctly updated the composer.json file but for this modif to be effective you also need to update the lock file by running `composer update prestashop/ps_apiresources` Currently you json and lock files are not in sync which is why all the CI are failing because we can't run `composer install`

File: composer.lock
Comment (jolelievre) [inline composer.lock:1]: The composer.lock contains too many updates, probably why all the CIs are failing The changes on this file should be reverted and then we need to run this command `composer update prestashop/ps_apiresources`

Upating the module only is critical or else all the dependencies are updated, and some updates cause unwanted side effects

File: src/Adapter/Module/CommandHandler/ResetModuleHandler.php
Comment (jolelievre) [inline src/Adapter/Module/CommandHandler/ResetModuleHandler.php]: The previous test was more accurate it tested both uninstalled and disabled states instead of both at the same time, we can also use the returned object from the method as the reference for this kind of checks instead of the module manager that hides the logic

The `Module` object already contains all the info we need so I think it should be our reference in all the CQRS handlers

Comment (jolelievre) [reply src/Adapter/Module/CommandHandler/ResetModuleHandler.php]:
```suggestion
        $module = $this->moduleRepository->getPresentModule($command->getTechnicalName()->getValue());

        if (!$module->isInstalled()) {
            throw new CannotResetModuleException('Cannot reset a module not installed', CannotResetModuleException::NOT_INSTALLED);
        }
        
        if (!$module->isActive()) {
            throw new CannotResetModuleException('Cannot reset a disabled module', CannotResetModuleException::NOT_ACTIVE);
        }
```

File: src/Adapter/Module/CommandHandler/UpdateModuleStatusHandler.php
Comment (jolelievre) [inline src/Adapter/Module/CommandHandler/UpdateModuleStatusHandler.php:47]: Perfect, this also should be added in the `GetModuleInfosHandler` so we have a uniformized way to get the module everywhere

File: src/Core/Module/ModuleRepository.php
Comment (jolelievre) [inline src/Core/Module/ModuleRepository.php]:
```suggestion
    /**
     * Returns an instance of a present module, if the module is not in the modules folder an exception is thrown.
     */
    public function getPresentModule(string $technicalName): ModuleInterface 
```

**Review by kpodemski — DISMISSED** [review]
Hello @jinpresta 

We recently introduced the `9.0.x` branch, which will be the base for PrestaShop 9.0. If you want to see your change in this version, please rebase your PR onto `9.0.x`. Branch `develop` will be used for PrestaShop 9.1, so you need to decide what to do.

Please let me know your decision so I can unblock the pull request.

**Review by jolelievre — DISMISSED** [review]

**Review by boherm — APPROVED** [review]

**Review by tleon — APPROVED** [review]

**Comment by ps-jarvis** [comment]
PR merged, well done!

Message to @PrestaShop/committers: do not forget to milestone it before the merge.

**Pull request merged**
